app/controllers/projects_controller.rb

```diff
@@ -241,91 +243,139 @@ def updateLikedStatus
-                                     unit: params["#{field.id}_unit"],
-                                     index: params["#{field.id}_index"],
-                                     restrictions: restrictions)
-        respond_to do |format|
-          flash[:error] = 'Field names must be unique.'
-          redirect_to "/projects/#{@project.id}/edit_fields" and return
-        end
-      end
-    end
+      # Add fields based on type
+      errors += add_location_field(params)
+      errors += add_timestamp_field(params)
+      errors += add_number_fields(params)
+      errors += add_text_fields(params)
 
-    # Add fields based on type
-    if params[:hidden_location_count] == '1'
-      if addField('Latitude', params[:latitude], 'deg', [], params['latitude_index']) == -1 and return
-      end
-      if addField('Longitude', params[:longitude], 'deg', [], params['longitude_index']) == -1 and return
+      # If there's recorded error messages, roll back the database
+      if errors.length != 0
+        success = false
+        fail ActiveRecord::Rollback
       end
     end
 
-    if params[:hidden_timestamp_count] == '1'
-      if addField('Timestamp', params[:timestamp], '', [], params['timestamp_index']) == -1 and return
-      end
+    # Don't recalculate anything if we ended up rolling back, because nothing changed
+    if success
+      @project.reload
+      @project.recalculate_data_sets
     end
 
-    (params[:hidden_num_count].to_i).times do |i|
-      if addField('Number', params[('number_' + (i + 1).to_s).to_sym], params[('units_' + (i + 1).to_s).to_sym], [],  params[('number_' + ((i + 1).to_s) + '_index').to_sym]) == -1 and return
+    # redirect_to "/projects/#{@project.id}", notice: 'Fields were successfully updated.'
+    respond_to do |format|
+      if success
+        response = { redirect: url_for(@project) }
+        format.json { render json: response,  status: :ok }
+      else
+        response = { errors: errors.uniq }
+        format.json { render json: response,  status: :unprocessable_entity }
       end
     end
+  end
 
-    (params[:hidden_text_count].to_i).times do |i|
-      # Need to explicitly check if restrictions are nil because empty restrictions should be []
-      restrictions = params[('restrictions_' + (i + 1).to_s).to_sym].nil? ? [] : params[('restrictions_' + (i + 1).to_s).to_sym].split(',')
+  def save_formula_fields
+    @project = Project.find(params[:id])
 
-      if addField('Text', params[('text_' + (i + 1).to_s).to_sym], '', restrictions, params[('text_' + (i + 1).to_s + '_index').to_sym]) == -1 and return
+    errors = []
+    success = true
+
+    ActiveRecord::Base.transaction do
+      delete_hidden_fields(FormulaField, params[:hidden_deleted_fields])
+
+      # Update existing fields, create restrictions if any exist
+      @project.formula_fields.each do |field|
+        errors += update_formula_field(field, params)
+        # do formula validation here
```

Comment:
It happens on the new line 355. `try_execute` runs a lex/parse, so any errors from formula validation will crop up there.